README.txt
Nagios plugin written in python to monitor Filemaker services. 

I've testing against Filemaker12.

If you encounter any problems with check_filemaker, need help understanding it, or would
like to offer suggestions or feedback, please get in touch with me:

* Ask a question through our web support interface, at 
  https://answers.launchpad.net/~tanner

* Report bugs at https://bugs.launchpad.net/~tanner


CHANGELOG.txt
2012-09-02  Robert J. Tanner  <tanner@real-time.com>

	First iteration of setup.py and associated files

2012-08-31  Robert J. Tanner  <tanner@real-time.com>

	Initial import



tests/test_checkfilemaker.py
# Copyright (C) 2012 Real Time Enterprises, Inc.
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
#
from nose import with_setup
from nose.tools import raises
from check_filemaker import CheckFilemaker
import ConfigParser, os

class TestCheckFilemaker:
    
    def setUp(self):
        config = ConfigParser.ConfigParser()
        config.readfp(open('mytesting.cfg'))
        
        self.hostname = config.get('testing', 'hostname')
        self.username = config.get('testing', 'username')
        self.password = config.get('testing', 'password')
        
        self.databases = config.get('testing', 'databases').split(',')
        self.databases = [x.strip(' ') for x in self.databases]
        
        self.layout = config.get('testing', 'layout')
        
        self.fm = CheckFilemaker()
        self.fm.connect(self.hostname, self.username, self.password, self.databases[0], self.layout)
            
    def tearDown(self):
       pass
       
    def test_1(self):
        """Make sure we can find the database on the filemaker server"""
        
        mydb = [self.databases[0]]
        ret = self.fm.check_databases(mydb)
        assert(ret == set(mydb))
    
    def test_1_1(self):
        """Make sure we can find multiple database on the filemaker server"""
        
        mydb = self.databases
        ret = self.fm.check_databases(mydb)
        assert(ret == set(mydb))
                    
    def test_2(self):
        """Make sure we can do NOT find bad_db on the filemaker server"""

        # Umm, make sure you don't have a datbase called bad_db!
        #
        mydb = ['bad_db']
        ret = self.fm.check_databases(mydb)
        assert(ret == set([]))
    
    def test_2_1(self):
        """Make sure we catch when a database is not on the filemaker server"""
        
        mydb = self.databases
        ret = self.fm.check_databases(mydb)
        assert(ret == set(self.databases))
        
        mydb.append('baddb')
        assert(ret != set(mydb))

    @raises(Exception)    
    def test_3(self):
        """Test that bad hostname throws an exception"""
        
        fm2 = CheckFilemaker()
        fm2.connect('badhost.company.tld', self.username, self.password, self.database[0])

        mydb = [self.database[0]]
        ret = fm2.check_databases(mydb)
        
    @raises(Exception)    
    def test_4(self):
        """Test that bad username throws an exception"""
        
        fm2 = CheckFilemaker()
        fm2.connect(self.hostname, 'baduser', self.password, self.database[0])
        
        mydb = [self.database[0]]
        ret = fm2.check_databases(mydb)
        # I think PyFileMaker is buggy! Should get some sort of error/exception for bad user?
        #
        assert(False)
        

    @raises(Exception)    
    def test_5(self):
        """Test that bad password throws an exception"""
        
        fm2 = CheckFilemaker()
        fm2.connect(self.hostname, self.username, 'badpassword', self.database[0])

        mydb = [self.database[0]]
        ret = fm2.check_databases(mydb)
        # I think PyFileMaker is buggy! Should get some sort of error/exception for bad user?
        #
        assert(False)
       
    def test_6(self):
        """Make sure we can find layout layout on the filemaker server"""

        # WARNING: layout names are case sensitive! sysSetting != syssetting
        layouts = [self.layout]
        ret = self.fm.check_layouts(layouts)
        assert(ret == set(layouts))
        
    def test_7(self):
        """Look for column named Constant with a value of 1"""

        # Might be specific to my testing Filemaker server
        #
        ret = self.fm.column(Constant='1')
        assert(len(ret) > 0)
        
    @raises(Exception)      
    def test_8(self):
        """Look for column named BadColumn with a value of 1"""

        ret = self.fm.column(BadColumn='1')

    @raises(Exception)            
    def test_9(self):
        """Look for column named Constant with a value of 2"""

        ret = self.fm.column(Constant='2')

check_filemaker.py
#!/usr/bin/env python
#
# Copyright (C) 2012 Real Time Enterprises, Inc.
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
#
import sys, os, argparse
import getopt
from PyFileMaker import FMServer, FMServerError, FMError
import ast

__here__ = os.path.abspath(os.path.dirname(__file__))
__version__ = open(os.path.join(__here__, 'version.txt')).read().strip()

__copyright__ = 'Copyright (C) 2012 Real Time Enterprises, Inc.'
__min_version__ = (2, 6)
__app_name__ = 'check_filemaker'
        
class CheckFilemaker:
    OK       = 0
    WARNING  = 1
    CRITICAL = 2
    UNKNOWN  = 3
    
    def connect(self, hostname, username, password, database, layout, debug = False):
        auth = username + ':' + password + '@' + hostname
        self._fm = FMServer(auth, database, layout)
        self._debug = debug
    
    def column(self, **params):
        return self._fm.doFind(params)
        
    def check_layouts(self, layouts):
        lays = self._fm.getLayoutNames()
        return set(lays) & set(layouts)
        
    def check_databases(self, databases):
        dbs = self._fm.getDbNames()
        return set(dbs) & set(databases)
        
def check_python():    
    if sys.version_info < __min_version__:
        print >> sys.stderr, 'Error: cannot find a suitable python interpreter'
        print >> sys.stderr, '  Need python %d.%d or later' % __min_version__
        print >> sys.stderr, '  Found python %d.%d' % (sys.version_info[0], sys.version_info[1])
        print >> sys.stderr, '  CRITICAL - Cannot find suitable python interpreter'
        raise SystemExit, CheckFilemaker.CRITICAL

def parse_cli(argv, release):
    
    parser = argparse.ArgumentParser(description='Nagios plugin to check status of Filemaker databases',
        epilog='Send email to tanner@real-time.com if you have questions regarding this software.')
    
    parser.add_argument('--version', action='version', version=release,
        help='print version information')
    parser.add_argument('-H', '--hostname', dest='hostname', required=True,
        help='Host name, IP Address')
    parser.add_argument('-d', '--database', dest='database', required=True,
        help='Check database with indicated name')
    parser.add_argument('-l', '--layout', dest='layout', required=True,
        help='Uses layout with indicated name')
    parser.add_argument('-c', '--column', dest='column', required=True, metavar='{"COLUMN":"VALUE"}',
        help='Look for {"COLUMN":"VALUE"} in the database')
    parser.add_argument('-u', '--username', dest='username', required=True,
        help='connect using the indicated username')
    parser.add_argument('-p', '--password', dest='password', required=True,
        help='use the indicated password to authenticate the connection')

    args = parser.parse_args()
    return args
       
if __name__ == "__main__":   
         
    try:     
        release = __app_name__ + ' ' + __version__

        check_python()
        args = parse_cli(sys.argv, release)

        fm = CheckFilemaker()
        
        hostname = args.hostname
        database = args.database
        layout = args.layout

        fm.connect(hostname, args.username, args.password, database, layout)
        
        dbs = [database]
        if fm.check_databases(dbs) != set(dbs):            
            raise Exception('Did not find the required database: ' + str(dbs))
            
        lays = [layout]
        if fm.check_layouts(lays) != set(lays):
            raise Exception('Did not find the required layout: ' + str(lays))
        
        # Literal evail of --column="{'Constant':'1'}" see 
        # http://stackoverflow.com/questions/988228/converting-a-string-to-dictionary
        #
        parm = ast.literal_eval(args.column)        
        column = fm.column(**parm)
        if len(column) <= 0:
            raise Exception('Did not find required column: ')
        
        print '%s OK - %s' % (release, database)    
        raise SystemExit, CheckFilemaker.OK
    except FMServerError as e:
        print 'CRITICAL - FMServerError: ', e
        raise SystemExit, CheckFilemaker.CRITICAL
    except FMError as e:
        print 'CRITICAL - FMError: ', e
        raise SystemExit, CheckFilemaker.CRITICAL
    except KeyError as e:
        print 'CRITICAL - Missing required command line parameter:', e
        raise SystemExit, CheckFilemaker.CRITICAL
    except Exception as e:
        print type(e)
        print 'CRITICAL - Error: ', e
        raise SystemExit, CheckFilemaker.CRITICAL
